File: 자료구조/족보/submit/hw/PP06/StackType.py
MAX_ITEMS = 10


class NodeType:
    """ Node Type """

    def __init__(self, item):
        self.info = item
        self.next = None


class StackType:
    def __init__(self):
        self.topPtr = None

    def is_full(self):
        try:
            location = NodeType("test")
            del location
            return False
        except:
            return True

    def is_empty(self):
        return self.topPtr is None

    def push(self, item):
        new_item = NodeType(item)
        new_item.next = self.topPtr
        self.topPtr = new_item

    def pop(self):
        target = self.topPtr
        self.topPtr = self.topPtr.next
        del target

    def top(self):
        if self.is_empty():
            print("Failed to Top")
        else:
            return self.topPtr.info

    def __str__(self):
        location = self.topPtr
        while location != None:
            print(location.info, end=" ")
            location = location.next

File: 자료구조/족보/LAB/submit/2017100782_허진석_LAB07/ex02/StackType.cpp
#include <cstddef>
#include <new>
#include <iostream>
#include "StackType.h"

void StackType::Push(ItemType newItem)
{
	if (IsFull())
		throw FullStack();
	else
	{
		NodeType *location;
		location = new NodeType;
		location->info = newItem;
		location->next = topPtr;
		topPtr = location;
	}
}
void StackType::Pop()
{
	if (IsEmpty())
		throw EmptyStack();
	else
	{
		NodeType *tempPtr;
		tempPtr = topPtr;
		topPtr = topPtr->next;
		delete tempPtr;
	}
}
ItemType StackType::Top()
{
	if (IsEmpty())
		throw EmptyStack();
	else
		return topPtr->info;
}
bool StackType::IsEmpty() const
{
	return (topPtr == NULL);
}

void StackType::MakeEmpty()
{
	NodeType *tempPtr;

	while (topPtr != NULL)
	{
		tempPtr = topPtr;
		topPtr = topPtr->next;
		delete tempPtr;
	}
}

bool StackType::IsFull() const
{
	NodeType *location;
	try
	{
		location = new NodeType;
		delete location;
		return false;
	}
	catch (std::bad_alloc)
	{
		return true;
	}
}

StackType::~StackType()
{
	MakeEmpty();
}

StackType::StackType() // Class constructor.
{
	topPtr = NULL;
}

void StackType::Copy(StackType& anotherStack) const
{
	StackType tmp;
	NodeType *tmpNode = topPtr;
	
	anotherStack.MakeEmpty();

	while (tmpNode != NULL) {
		tmp.Push(tmpNode->info);
		tmpNode = tmpNode->next;
	}
	while (!tmp.IsEmpty()) {
		anotherStack.Push(tmp.Top());
		tmp.Pop();
	}
}

void StackType::Print()
{
	NodeType *tmp = topPtr;
	while (tmp != NULL)
	{
		std::cout << tmp->info << " ";
		tmp = tmp->next;
	}
	std::cout << std::endl;
}

StackType& StackType::operator= (const StackType& anotherStack) {
	anotherStack.Copy(*this);

	return (*this);
}
